### PR TITLE
[Declarative config] Phase 3: Parse the full configuration document

### PR DESCRIPTION
**Pull request by stevejgordon**
Contibutes to #7658

## Changes

Parse the entire declarative configuration document, including sections that are not yet applied. This ensures invalid YAML and environment variable substitution issues are reported instead of being silently ignored.

Expands test coverage for aliases, merge keys, substitutions, malformed documents, and other YAML edge cases.

Also optimises EventListener tests to avoid long test execution times.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] ~Changes in public API reviewed (if applicable)~

### CODE REVIEW

File: test/OpenTelemetry.Configuration.Declarative.FuzzTests/DeclarativeConfigurationFuzzTests.cs [hunk 1]
```diff
@@ -46,4 +77,249 @@ public Property LoadArbitraryYamlNeverThrowsUnexpectedException() =>
...
+            from children in Gen.ArrayOf(childGenerator, childCount)
+            select (YamlValue)new SequenceValue(hasAnchor, children);
+
+        return Gen.Frequency(
+            (3, ScalarOrAliasGenerator),
+            (2, mappingGenerator),
+            (2, sequenceGenerator));
+    }
+
+    // Range/index syntax is unavailable on net462, which this project targets.
```
Comment (martincostello) [inline test/OpenTelemetry.Configuration.Declarative.FuzzTests/DeclarativeConfigurationFuzzTests.cs]:
```suggestion
```

File: src/OpenTelemetry.Configuration.Declarative/Internal/Yaml/YamlPath.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    /// <returns>The composed path.</returns>
+    internal static string Child(string parent, string key) => $"{parent}.{key}";
+
+    /// <summary>
+    /// Returns the path of the item at <paramref name="index"/> of <paramref name="parent"/>.
+    /// </summary>
+    /// <param name="parent">The path of the sequence that holds the item.</param>
+    /// <param name="index">The zero-based item index.</param>
+    /// <returns>The composed path.</returns>
+    internal static string Index(string parent, int index) => $"{parent}[{index}]";
```
Comment (martincostello) [inline src/OpenTelemetry.Configuration.Declarative/Internal/Yaml/YamlPath.cs]: Should make this explicitly invariant.

File: src/OpenTelemetry.Configuration.Declarative/Internal/Yaml/YamlNodeReferenceEqualityComparer.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    /// </summary>
+    internal static readonly YamlNodeReferenceEqualityComparer Instance = new();
+
+    private YamlNodeReferenceEqualityComparer()
+    {
+    }
+
+    bool IEqualityComparer<YamlNode>.Equals(YamlNode? x, YamlNode? y) => ReferenceEquals(x, y);
+
+    int IEqualityComparer<YamlNode>.GetHashCode(YamlNode obj) => RuntimeHelpers.GetHashCode(obj);
```
Comment (martincostello) [inline src/OpenTelemetry.Configuration.Declarative/Internal/Yaml/YamlNodeReferenceEqualityComparer.cs:30]: How come these are explicitly implemented?

File: src/OpenTelemetry.Configuration.Declarative/Internal/Yaml/YamlMergeKeyValidator.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    private static bool IsMergeKey(YamlNode key) =>
+        key is YamlScalarNode scalar &&
+        ((!scalar.Tag.IsEmpty &&
+                !scalar.Tag.IsNonSpecific &&
+                string.Equals(scalar.Tag.Value, MergeTag, StringComparison.Ordinal)) ||
+            (string.Equals(scalar.Value, MergeKey, StringComparison.Ordinal) &&
+                YamlScalarResolver.IsPlain(scalar) &&
+                scalar.Tag.IsEmpty));
```
Comment (martincostello) [inline src/OpenTelemetry.Configuration.Declarative/Internal/Yaml/YamlMergeKeyValidator.cs]: The indentation here isn't very readable IMHO.

File: src/OpenTelemetry.Configuration.Declarative/Internal/Yaml/YamlMergeKeyValidator.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    private static void Visit(YamlNode node, string path, HashSet<YamlNode> visited)
+    {
+        if (!visited.Add(node))
+        {
+            return;
+        }
+
+        switch (node)
```
Comment (martincostello) [inline src/OpenTelemetry.Configuration.Declarative/Internal/Yaml/YamlMergeKeyValidator.cs:57]: Please add a `default` case to avoid a warning in #6899.

File: src/OpenTelemetry.Configuration.Declarative/Internal/Yaml/YamlNodeReferenceEqualityComparer.cs (hunk 3 above)
Comment (stevejgordon) [reply src/OpenTelemetry.Configuration.Declarative/Internal/Yaml/YamlNodeReferenceEqualityComparer.cs:30]: This was deliberate. The only intended API is `Instance` so I wanted to avoid the `Equals` and `GetHashCode` methods appearing on the class itself. They belong to the `IEqualityComparer<YamlNode>` contract. Happy to switch to implicit if you prefer the simpler form, though.

File: src/OpenTelemetry.Configuration.Declarative/Internal/Yaml/YamlMergeKeyValidator.cs (hunk 5 above)
Comment (stevejgordon) [reply src/OpenTelemetry.Configuration.Declarative/Internal/Yaml/YamlMergeKeyValidator.cs:57]: Addressed in d1a0aba6f6ccfdfdf1b0bcfefc0d65f72874462c

File: src/OpenTelemetry.Configuration.Declarative/Internal/Yaml/YamlMergeKeyValidator.cs (hunk 4 above)
Comment (stevejgordon) [reply src/OpenTelemetry.Configuration.Declarative/Internal/Yaml/YamlMergeKeyValidator.cs]: Addressed in d1a0aba6f6ccfdfdf1b0bcfefc0d65f72874462c

File: src/OpenTelemetry.Configuration.Declarative/Internal/Yaml/YamlPath.cs (hunk 2 above)
Comment (stevejgordon) [reply src/OpenTelemetry.Configuration.Declarative/Internal/Yaml/YamlPath.cs]: Addressed in d1a0aba6f6ccfdfdf1b0bcfefc0d65f72874462c

File: test/OpenTelemetry.Configuration.Declarative.FuzzTests/DeclarativeConfigurationFuzzTests.cs (hunk 1 above)
Comment (stevejgordon) [reply test/OpenTelemetry.Configuration.Declarative.FuzzTests/DeclarativeConfigurationFuzzTests.cs]: Addressed in d1a0aba6f6ccfdfdf1b0bcfefc0d65f72874462c - I remove the helper entirely in favour of LINQ which is cross TFM compatibale and doesn't require explanation.

**Review by martincostello — APPROVED** [review]
